metadata/src/from_into/v14.rs

```diff
@@ -178,17 +226,375 @@ fn v14_to_v15(metadata: v14::RuntimeMetadataV14) -> v15::RuntimeMetadataV15 {
+            let Some(ident) = ty.ty.path.ident() else { return false };
+            ident == "RuntimeCall"
+        })
+        .expect("RuntimeCall exists in V14; qed");
+
+    let event_enum = metadata
+        .types
+        .types
+        .iter()
+        .find(|ty| {
+            let Some(ident) = ty.ty.path.ident() else { return false };
+            ident == "RuntimeEvent"
+        })
+        .expect("RuntimeEvent exists in V14; qed");
+
+    let call_ty = call_enum.id.into();
+    let event_ty = event_enum.id.into();
+
+    let mut path_segments = call_enum.ty.path.segments.clone();
+    let last = path_segments
+        .last_mut()
+        .expect("Should have at least one segment checked above; qed");
+    *last = "RuntimeError".to_string();
+
+    let error_ty_id = generate_runtime_error_type(metadata, path_segments);
+
+    v15::OuterEnums {
+        call_enum_ty: call_ty,
+        event_enum_ty: event_ty,
+        error_enum_ty: error_ty_id.into(),
+    }
+}
+
+/// Generate the `RuntimeError` type and add it to the metadata.
+///
+/// Returns the `RuntimeError` Id from the registry.
+fn generate_runtime_error_type(
+    metadata: &mut v14::RuntimeMetadataV14,
+    path_segments: Vec<String>,
+) -> u32 {
+    let variants: Vec<_> = metadata
+        .pallets
+        .iter()
+        .filter_map(|pallet| {
+            let Some(pallet_error) = &pallet.error else { return None };
+
+            let error_ty = metadata
+                .types
+                .resolve(pallet_error.ty.id)
+                .expect("Error type is present in the metadata; qed");
+
+            // Note: the type name is the name of the type as it appears in substrate, we cannot
+            // accurately reproduce it when it contains more than 1 generic.
+            // The first generic is always the `Runtime`, the others represent instances.
+            let instances = ",Instance".repeat(error_ty.type_params.len() - 1);
+            let path = format!("{}<Runtime{}>", error_ty.path, instances);
+
+            Some(scale_info::Variant {
+                name: pallet.name.clone(),
+                fields: vec![scale_info::Field {
+                    name: None,
+                    ty: pallet_error.ty.id.into(),
+                    type_name: Some(path),
+                    docs: vec![],
+                }],
+                index: pallet.index,
+                docs: vec![],
+            })
+        })
+        .collect();
+
+    let error_type = scale_info::Type {
+        path: scale_info::Path {
+            segments: path_segments,
+        },
+        type_params: vec![],
+        type_def: scale_info::TypeDef::Variant(scale_info::TypeDefVariant { variants }),
+        docs: vec![],
+    };
+
+    let error_type_id = metadata.types.types.len() as u32;
+
+    metadata.types.types.push(scale_info::PortableType {
+        id: error_type_id,
+        ty: error_type,
+    });
+
+    error_type_id
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Like the tests; converting to v14 and back to v15 again and checking that things line up still is a good way to check :)